Refactor NT offsets metadata to be a Tensor

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96354
* __->__ #96909

It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

Bonus:
* Remove `_tensor` suffixes from metadata / getter names

jbschlosser committed:
Refactor NT offsets metadata to be a Tensor

[ghstack-poisoned]

drisspg commented on aten/src/ATen/NestedTensorImpl.cpp:
```diff
@@ -11,21 +11,22 @@
 
 #include <numeric>
 #include <functional>
+#include <iostream>
```
Looks like somone is trying to hitch a ride

drisspg commented on aten/src/ATen/NestedTensorImpl.cpp:
```diff
@@ -167,11 +169,11 @@ NestedTensorImpl::NestedTensorImpl(
       nested_size_tensor_(std::move(nested_size_tensor)),
       nested_stride_tensor_(std::move(nested_stride_tensor)),
-      storage_offsets_(std::move(offsets)),
+      offsets_tensor_(std::move(offsets_tensor)),
```
Nit: I think now that everything is a tensor baking the type into the name is ugly ( I think this was partially my doing).  I like nested_size_, nested_stride_, storage_offsets_. I understand not wanting to do this and just observation

jbschlosser replied:
I do agree with you; I'm going back and forth on whether I want to go through with it though :p

jbschlosser replied:
Okay I did it

drisspg commented on aten/src/ATen/NestedTensorImpl.cpp:
```diff
@@ -224,13 +226,13 @@ NestedTensorImpl::NestedTensorImpl(
     at::Tensor nested_size_tensor,
     at::Tensor nested_stride_tensor,
-    std::vector<int64_t>&& offsets)
+    at::Tensor offsets_tensor)
```
Other random comment: I still want a cpp expert to infer if we should taking in rvalue references for all pieces of metadata. Alban said that moving the temps will avoid refcount bumps and since we are (for now) assuming the callers of the constructor will create new unique pieces of metadata then we should be using rvalues.

But still clang tidy will say: `Moving a temporary object prevents copy elision (fix available)` So idk. PREMATURE OPTIMIZATION FOR THE WIN!

jbschlosser replied:
Did we determine this is fine or does something need to change?

drisspg replied:
This is goood

drisspg commented on aten/src/ATen/native/nested/NestedTensorBinaryOps.cpp:
```diff
@@ -60,11 +60,12 @@ get_elementwise_nested_tensor_impl(
   bool offsets_match = true;
-  for (size_t i = 0; i < self_offsets.size(); i++) {
-    offsets_match = offsets_match && (self_offsets[i] == other_offsets[i]);
+  for (size_t i = 0; i < self_offsets.size(0); i++) {
```
could also you at::equal? but very likely more efficient to do the ptr loop

drisspg reviewed (approved):
Overall looks good, I don't have a sense of how much "heavier" an at::Tensor of longs is verse the std::Vector of longs. And I am still a fan of storage_offsets as a name since I think it is more explicit but good job, this must have sucked.

ezyang commented:
Tensor is substantially heavier than a vector of longs, in terms of both memory usage as well as ref counting. However, all of this should be a non-issue if you get substantial amounts of fusion with PT2

jbschlosser committed:
Update on "Refactor NT offsets metadata to be a Tensor"


It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

[ghstack-poisoned]

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -110,7 +110,7 @@ def test_2d_nested_tensor(self, batch_size, max_seq_len, vocab_size):
             row = list(np.random.randint(low=0, high=vocab_size, size=(length,)))
             data.append(row)
-            nested_tensor_ref_list.append(torch.tensor(row))
+            nested_tensor_ref_list.append(torch.Tensor(row))
```
not sure why tf this just started failing?

drisspg replied:
thats weird and not good lol

jbschlosser committed:
Update on "Refactor NT offsets metadata to be a Tensor"


It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

Bonus:
* Remove `_tensor` suffixes from metadata / getter names

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5774,10 +5774,11 @@
   autogen: _nested_tensor_strides.out
 
-- func: _nested_tensor_offsets(Tensor self) -> int[]
+- func: _nested_tensor_storage_offsets(Tensor self) -> Tensor
```
Changing the name of a native function worries me

jbschlosser replied:
As discussed, we had a manual exception added to the gen code before so this wasn't exposed to Python until now.

Unclear whether the FC / BC check will fail due to the name change (it shouldn't, but idk if it's aware of the gen code exception). Guess we'll see

jbschlosser committed:
Update on "Refactor NT offsets metadata to be a Tensor"


It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

Bonus:
* Remove `_tensor` suffixes from metadata / getter names

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Refactor NT offsets metadata to be a Tensor"


It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

Bonus:
* Remove `_tensor` suffixes from metadata / getter names

[ghstack-poisoned]

jbschlosser committed:
Update on "Refactor NT offsets metadata to be a Tensor"


It's tedious work, but somebody's gotta do it.

Benefits:
* Enable access to offsets metadata from Python via private API (for validation, etc.)
* Consistency with nested sizes / strides metadata
* Needed for SymInt-ifying offsets metadata
* more TBD

Bonus:
* Remove `_tensor` suffixes from metadata / getter names

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
